src/main/java/com/commercetools/project/sync/product/service/impl/ProductReferenceTransformServiceImpl.java

```diff
@@ -0,0 +1,467 @@
+package com.commercetools.project.sync.product.service.impl;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+import com.commercetools.project.sync.model.ResourceIdsGraphQlRequest;
+import com.commercetools.project.sync.product.service.ProductReferenceTransformService;
+import com.commercetools.project.sync.service.impl.BaseServiceImpl;
+import com.commercetools.project.sync.util.referenceresolution.ProductReferenceResolutionUtils;
+import com.commercetools.sync.commons.models.GraphQlQueryResources;
+import com.commercetools.sync.commons.models.ResourceKeyId;
+import com.commercetools.sync.commons.utils.ChunkUtils;
+import com.commercetools.sync.customobjects.helpers.CustomObjectCompositeIdentifier;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.channels.Channel;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.customobjects.queries.CustomObjectQuery;
+import io.sphere.sdk.models.Reference;
+import io.sphere.sdk.products.Price;
+import io.sphere.sdk.products.Product;
+import io.sphere.sdk.products.ProductDraft;
+import io.sphere.sdk.products.ProductLike;
+import io.sphere.sdk.products.ProductVariant;
+import io.sphere.sdk.producttypes.ProductType;
+import io.sphere.sdk.states.State;
+import io.sphere.sdk.taxcategories.TaxCategory;
+import io.sphere.sdk.types.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+
+public class ProductReferenceTransformServiceImpl extends BaseServiceImpl
+    implements ProductReferenceTransformService {
+  private final Map<String, String> allResourcesIdToKey = new HashMap<>();
+  /*
+   * An id is a 36 characters long string. (i.e: 53c4a8b4-754f-4b95-b6f2-3e1e70e3d0c3) We
+   * chunk them in 300 ids, we will have a query around 11.000 characters. Above this size it
+   * could return - Error 413 (Request Entity Too Large)
+   */
+  public static final int CHUNK_SIZE = 300;
+
+  public ProductReferenceTransformServiceImpl(@Nonnull final SphereClient ctpClient) {
+    super(ctpClient);
+  }
+
+  @Nonnull
+  @Override
+  public CompletionStage<List<ProductDraft>> transformProductReferences(
```

Comment:
Note: Most of the concerns I have commented on in CategoryService applies to here, I am not adding the same comments to not overload PR.

Reply:
https://github.com/commercetools/commercetools-project-sync/pull/240/commits/7e34c3b17e1a83cf15294dd9044e07c8138f5562 , https://github.com/commercetools/commercetools-project-sync/pull/240/commits/7918a4f878936800758c9e167c0d5f6287f711a9